File: csharp/FunctionalRefactoring.Tests/AppTests.cs
using FunctionalRefactoring.Models;
using Xunit;

namespace FunctionalRefactoring.Tests
{
    public class AppTests
    {
        [Fact]
        public void HappyPath()
        {
            var cartId = new CartId("some-gold-cart");
            var storage = new SpyStorage();

            App.ApplyDiscount(cartId, storage);

            var expected = new Cart(new CartId("some-gold-cart"), new CustomerId("gold-customer"), new Amount(50));

            Assert.Equal(expected, storage.Saved);
        }

        [Fact]
        public void NoDiscount()
        {
            var cartId = new CartId("some-normal-cart");
            var storage = new SpyStorage();

            App.ApplyDiscount(cartId, storage);

            Assert.Null(storage.Saved);
        }

        [Fact]
        public void MissingCart()
        {
            var cartId = new CartId("missing-cart");
            var storage = new SpyStorage();

            App.ApplyDiscount(cartId, storage);

            Assert.Null(storage.Saved);
        }

        class SpyStorage : IStorage<Cart>
        {
            public Cart Saved { get; private set; }

            public void Flush(Cart item)
            {
                Saved = item;
            }
        }
    }
}

File: scala/src/test/scala/org/fprefactoring/AppTests.scala
package org.fprefactoring

import org.fprefactoring.App._
import org.fprefactoring.Models._
import org.scalatest.FunSuite

class AppTests extends FunSuite {

  test("happy path") {
    val cartId = CartId("some-gold-cart")
    val storage = new SpyStorage

    applyDiscount(cartId, storage)

    val expected = Cart(CartId("some-gold-cart"), CustomerId("gold-customer"), Amount(50))
    assert(storage.saved.get == expected)
  }

  test("no discount") {
    val cartId = CartId("some-normal-cart")
    val storage = new SpyStorage

    applyDiscount(cartId, storage)

    assert(storage.saved.isEmpty)
  }

  test("missing cart") {
    val cartId = CartId("missing-cart")
    val storage = new SpyStorage

    applyDiscount(cartId, storage)

    assert(storage.saved.isEmpty)
  }

  class SpyStorage extends Storage[Cart] {
    var saved: Option[Cart] = None
    override def flush(value: Cart): Unit = saved = Some(value)
  }
}
